exercise4-1/solution/system_nodes.cc

```diff
@@ -0,0 +1,43 @@
+#include "system_nodes.h"
+
+#include <cactus_rt/utils.h>
+
+#include <string>
+
+#include "utils.h"
+
+using camera_demo_interfaces::msg::FakeImage;
+
+ImagePublisherNode::ImagePublisherNode(
+  std::shared_ptr<cactus_rt::tracing::ThreadTracer> tracer,
+  double frequency_hz
+) : Node("imagepub"), tracer_(tracer) {
+  timer_ = this->create_wall_timer(
+    std::chrono::microseconds(static_cast<int>(1000000 / frequency_hz)),
+    std::bind(&ImagePublisherNode::TimerCallback, this)
+  );
+
+  publisher_ = this->create_publisher<FakeImage>("/image", 10);
+  last_image_publisher_timepoint = 0;
+}
+
+void ImagePublisherNode::TimerCallback() {
+  FakeImage img;
+  img.data = {127};
+  img.published_at_monotonic_nanos = cactus_rt::NowNs();
+
+  publisher_->publish(img);
+
+  // show publisher period
+  // omit first publisher call
+  if (last_image_publisher_timepoint != 0){
+    tracer_->StartSpan("period", nullptr, last_image_publisher_timepoint);
+    tracer_->EndSpan(img.published_at_monotonic_nanos);
```

Comment:
Maybe instead of calling this `period`, call this the `Sleep`?

Reply:
I chose period, because the trace shows the period of the publisher. Why do you think, that "Sleep" is more appropriate?

Reply:
![image](https://github.com/ros-realtime/roscon-2023-realtime-workshop/assets/338100/fa9be3ee-f476-469b-a542-ee8a4673599d)

This is a screenshot from exercise 1. The `Sleep` is the time between the last iteration to the current iteration. I think in this case the sleep time == period time because the timer callback's execution duration is negligible. I suppose you can leave it as is as it will make the histogram work better.